cmd/lncli/walletrpc_active.go

```diff
@@ -726,6 +734,339 @@ type fundPsbtResponse struct {
 	Locks             []*utxoLease `json:"locks"`
 }
 
+var fundTemplatePsbtCommand = cli.Command{
+	Name: "fundtemplate",
+	Usage: "Fund a Partially Signed Bitcoin Transaction (PSBT) from a " +
+		"template.",
+	ArgsUsage: "[--template_psbt=T | [--outputs=O [--inputs=I]]] " +
+		"[--conf_target=C | --sat_per_vbyte=S] " +
+		"[--change_type=A] [--change_output_index=I]",
+	Description: `
+	The fund command creates a fully populated PSBT that contains enough
+	inputs to fund the outputs specified in either the template.
+
+	The main difference to the 'fund' command is that the template PSBT
+	is allowed to already contain both inputs and outputs and coin selection
+	and fee estimation is still performed.
+
+	If '--inputs' and '--outputs' are provided instead of a template, then
+	those are used to create a new PSBT template.
+
+	The 'outputs' flag decodes addresses and the amount to send respectively
+	in the following JSON format:
+
+	    --outputs='["ExampleAddr:NumCoinsInSatoshis", "SecondAddr:Sats"]'
+
+	The 'outputs' format is different from the 'fund' command as the order
+	is important for being able to specify the change output index, so an
+	array is used rather than a map.
+
+	After verifying and possibly adding new inputs, all input UTXOs added by
+	the command are locked with an internal app ID. Inputs already present
+	in the template are NOT locked.
```

Comment:
why not? Is there a way for a user to manually go lock the output then instead? 

cause it seems dangerous to start a psbt round with an input that LND might sweep out from under you for something else if we dont lock them

Reply:
also - this is different from the behaviour of the `fund` command.

Reply:
One other thing I noticed is: For the `fund` command, if I specify an input - it get's locked. Then if i try use that same input in another `fund` round - it fails due to `input 0 not found in list of non-locked UTXO` which seems good as this protects me. 

But with `fundtemplate` - im able to specify inputs of mine that are already locked. This seems dangerous. 

I get that we cant lock all the inputs since we dont necessarily own all the inputs, but maybe we should lock the ones that we do own?

Reply:
So the locking here is mostly to avoid an input specified in the template to be selected again by the coin selection algorithm. And also to make sure no _other_ application snatches the input away from us while we're using it.

So to your first question "why not" in the CLI description: Because we want the inputs specified in the template to _already_ be locked by the user. We didn't enforce that previously, but since that's dangerous, I now added that check.

> But with `fundtemplate` - im able to specify inputs of mine that are already locked. This seems dangerous.

Hmm, I'm not sure there's a lot we can do about that. We want the user to go ahead and select an input, manually lock it and then call `fundtemplate`. But we cannot really determine whether the user locked them manually or they are locked from a previous attempt. We could ask the user to provide the lock ID they used to lock them, to make sure they did it themselves. But then they can still just provide the lock ID of another process, as you can read the lock ID from the output of `lncli wallet listleases`.
So I think users concerned about using this RPC correctly would hopefully read the instructions and do the locking properly?
Or do you have an idea for how to make things even more explicit?

Reply:
> We could ask the user to provide the lock ID they used to lock them, to make sure they did it themselves

Jumping on a moving train here, so apologies if my comments are completely off-balance. But as a general UX rule, we should try to minimize the inputs user must provide.